optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/DeepCloningUtils.java

```diff
@@ -181,4 +187,56 @@ public boolean isFieldADeepCloneProperty(Field field, Class<?> owningClass) {
         }
         return false;
     }
+
+    /**
+     * @return never null
+     */
+    public Set<Class<?>> getDeepClonedTypeArguments(Type genericType) {
```

Comment:
This method (and others in this class) create a whole lot of collections just to throw them away a little while later.
All of that is going to create quite some GC pressure, and that is going to affect performance.
Of course, this is only a problem if these methods are called frequently on the hot path.

If that is the case, I would consider breaking a guideline on stateless methods and use one collection throughout the recursion. Alternatively, you can do this by concatenating streams, but that has performance downsides of its own.